File: 1285-balance-a-binary-search-tree/balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> order;
    void inor(TreeNode* node){
        if(node == NULL) return;
        inor(node->left);
        order.push_back(node->val);
        inor(node->right);
    }

    TreeNode* construct(int left, int right){
        if(left > right) return NULL;
        int mid = (left+right)/2;
        TreeNode* node = new TreeNode(order[mid]);
        node->left = construct(left, mid-1);
        node->right = construct(mid+1, right);
        return node;
    }

    TreeNode* balanceBST(TreeNode* root) {
        inor(root);
        return construct(0, order.size()-1);
    }
};

File: 1285-balance-a-binary-search-tree/balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def balanceBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        inor = []
        def inorder(root):
            if not root:
                return
            
            inorder(root.left)
            inor.append(root.val)
            inorder(root.right)
        
        def construct(left, right):
            if left > right:
                return None
            
            mid = (left+right)//2
            node = TreeNode(inor[mid])
            node.left = construct(left,mid-1)
            node.right = construct(mid+1, right)
            return node
        
        inorder(root)
        return construct(0,len(inor)-1)
        